### ISSUE TITLE
different nan handling in to_json() vs to_dict()

### ISSUE THREAD

**Issue by iuiu34** [pandas-dev/pandas#58230]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import json
df = pd.DataFrame(
        [{'a': 1, 'b': 1},  {'a': 2, 'b': None}]
    )
assert df.to_dict(orient='records') == json.loads(df.to_json(orient='records'))
# AssertionError
```

### Issue Description

I want to convert a dataFrame to a list of records with nulls as None (not nan, nor inf)

Using `df.to_dict(orient='records')' i can't get that. There's no option to customize nan parsing.
in the other side, in to_json(), nan parsing is done native (and works lovely, i have to say)

so, is desired behaviour, to use `json.loads(df.to_json(orient='records'))` to get my list? 
shouldn't be done in `.to_dict()`?

thanks

### Expected Behavior

assert df.to_dict(orient='records') == json.loads(df.to_json(orient='records'))
# True

### Installed Versions

INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.9.9.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19043
machine               : AMD64
processor             : Intel64 Family 6 Model 141 Stepping 1, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : es_ES.cp1252
pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 69.1.1
pip                   : 24.0
Cython                : None
pytest                : 8.1.1
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : 15.0.1
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.12.0
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

**Comment by Aloqeely** [comment]
I agree things should be consistent. what's even weirder with `to_dict` is that `pd.NA` gets converted to `None`, whereas `np.nan` and `None` are both converted to `nan`.

**Comment by asishm** [comment]
> what's even weirder with to_dict is that pd.NA gets converted to None, whereas np.nan and None are both converted to nan

`None` doesn't get converted to `nan` in the `to_json` step. Rather, it happens in the dataframe constructor. If you craft a dataframe that explicitly has`None`, you'll see that it does return a `None`.

```py
In [37]: pd.DataFrame({'a': [1, None, pd.NaT, np.nan, pd.NA]}, dtype='object').to_dict(orient='records')
Out[37]: [{'a': 1}, {'a': None}, {'a': NaT}, {'a': nan}, {'a': None}]
```

**Comment by rminkler1** [comment]
I'm seeing a lot of inconsistencies here, and they don't all appear to be pandas related.
Is this a bug, or intentional for some reason?
I'm suspecting this is intentional because `data[0]['b']` is an `Int`, `data[1]['b']`  as `None` is treated as numerical or `NaN`.
With all the json conversions as well, I wouldn't expect, `NaN`, `Null`, or `None` to be maintained.
While it would be ideal for them to match, I'm beginning to thing that won't be possible.

Test:
```
import pandas as pd
import json

data = [{'a': 1, 'b': 1},  {'a': 2, 'b': None}]
df = pd.DataFrame(
        data
    )
print("input data")
print(data)
print("\nafter conversion to dataframe")
print(df)
print("\nafter conversion to_dict")
print(df.to_dict(orient='records'))
print("\nafter conversion to_json")
print(df.to_json(orient='records'))
print("\nafter python json processing")
print(json.loads(df.to_json(orient='records')))
```
output:
```
input data
[{'a': 1, 'b': 1}, {'a': 2, 'b': None}]

after conversion to dataframe
   a    b
0  1  1.0
1  2  NaN

after conversion to_dict
[{'a': 1, 'b': 1.0}, {'a': 2, 'b': nan}]

after conversion to_json
[{"a":1,"b":1.0},{"a":2,"b":null}]

after python json processing
[{'a': 1, 'b': 1.0}, {'a': 2, 'b': None}]
```

**Comment by rminkler1** [comment]
Maybe `NaN` in Pandas should convert to `None` with `to_dict` instead of `nan`?
